### PR TITLE
UCP/WIREUP/CORE: Fix CM wireup fallback

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Fix CM wireup fallback.

## Why ?

To allow selecting multiple paths (if their are supported) in CM initial configuration. It was restricted by `AM_LANE_ONLY` flag 
during fallback, because `rc` and `dc` try to create CM configuration with auxiliary UD for keepalive.

## How ?

1. Introduce `UCP_EP_INIT_DISALLOW_AUX_TL_RSC` EP init flag to disallow selection of AUX transports at all.
2. Change order in EP init flags to have `UCP_EP_INIT_DISALLOW_AUX_TL_RSC` before `UCP_EP_INIT_CREATE_AM_LANE_ONLY` -  it is needed to prefer `UCP_EP_INIT_DISALLOW_AUX_TL_RSC` over `UCP_EP_INIT_CREATE_AM_LANE_ONLY` in `ucs_for_each_submask`.
3. Update `ucp_cm_client_uct_connect_progress` to do the following retries of initializing of CM configuration:
- `ep_init_flags` = `0`
- `ep_init_flags` = `UCP_EP_INIT_DISALLOW_AUX_TL_RSC`
- `ep_init_flags` = `UCP_EP_INIT_CREATE_AM_LANE_ONLY`
- `ep_init_flags` = `UCP_EP_INIT_DISALLOW_AUX_TL_RSC | UCP_EP_INIT_CREATE_AM_LANE_ONLY`.

### CODE REVIEW

File: src/ucp/core/ucp_ep.h [hunk 1]
```diff
@@ -159,9 +159,12 @@ enum {
     UCP_EP_INIT_CONNECT_TO_IFACE_ONLY  = UCS_BIT(7),  /**< Select transports which
                                                            support CONNECT_TO_IFACE
                                                            mode only */
-    UCP_EP_INIT_CREATE_AM_LANE_ONLY    = UCS_BIT(8),  /**< Endpoint requires an AM lane only */
-    UCP_EP_INIT_ALLOW_AUX_TL_RSC       = UCS_BIT(9)   /**< Endpoint allows selecting of auxiliary
-                                                           transports for its configuration */
+    UCP_EP_INIT_DISALLOW_AUX_TL_RSC    = UCS_BIT(8),  /**< Endpoint disallows selecting of auxiliary
```
Comment (brminich) [inline src/ucp/core/ucp_ep.h:162]: can we instead assume that if `UCP_EP_INIT_ALLOW_AUX_TL_RSC` is not present in the flag, than using aux is prohibited? (and update all places where it needs to be set)

Comment (brminich) [reply src/ucp/core/ucp_ep.h:162]: discussed offline, can still use 2 flags, but with different meaning:
 - flag allowing usage of aux transport for AM lane
 - flag allowing usage of aux transport for KA lane

Comment (dmitrygx) [reply src/ucp/core/ucp_ep.h]: discussed offline

File: src/ucp/wireup/select.c [hunk 2]
```diff
@@ -408,8 +408,7 @@ static UCS_F_NOINLINE ucs_status_t ucp_wireup_select_transport(
                 &context->tl_mds[context->tl_rscs[rsc_index].md_index].attr;
 
         if ((context->tl_rscs[rsc_index].flags & UCP_TL_RSC_FLAG_AUX) &&
-            !(criteria->tl_rsc_flags & UCP_TL_RSC_FLAG_AUX) &&
-            !(select_params->ep_init_flags & UCP_EP_INIT_ALLOW_AUX_TL_RSC)) {
+            !(criteria->tl_rsc_flags & UCP_TL_RSC_FLAG_AUX)) {
```
Comment (brminich) [inline src/ucp/wireup/select.c:411]: 👍

File: src/ucp/wireup/select.c [hunk 3]
```diff
@@ -1623,7 +1627,7 @@ ucp_wireup_keepalive_score_func(const ucp_worker_iface_t *wiface,
     }
 
     return ucp_wireup_am_score_func(wiface, md_attr, remote_iface_attr) *
-           (perf_attr.max_inflight_eps / SIZE_MAX);
+           ((double)perf_attr.max_inflight_eps / (double)SIZE_MAX);
```
Comment (brminich) [inline src/ucp/wireup/select.c:1630]: unrelated?

File: src/ucp/wireup/select.c [hunk 4]
```diff
@@ -1654,9 +1658,10 @@ ucp_wireup_add_keepalive_lane(const ucp_wireup_select_params_t *select_params,
     criteria.is_keepalive       = 1;
     criteria.calc_score         = ucp_wireup_keepalive_score_func;
     /* Keepalive can also use auxiliary transports */
-    criteria.tl_rsc_flags       = UCP_TL_RSC_FLAG_AUX;
+    criteria.tl_rsc_flags       =
+            (ep_init_flags & UCP_EP_INIT_KA_ALLOW_AUX_TL_RSC) ?
+            UCP_TL_RSC_FLAG_AUX : 0;
     ucp_wireup_fill_peer_err_criteria(&criteria, ep_init_flags);
-
```
Comment (brminich) [inline src/ucp/wireup/select.c:1659]: imo, no need to remove

File: src/ucp/wireup/wireup.c [hunk 5]
```diff
@@ -729,7 +731,8 @@ ucp_wireup_send_ep_removed(ucp_worker_h worker, const ucp_wireup_msg_t *msg,
                              UCP_EP_INIT_CONNECT_TO_IFACE_ONLY |
                              UCP_EP_INIT_CREATE_AM_LANE |
                              UCP_EP_INIT_CREATE_AM_LANE_ONLY |
-                             UCP_EP_INIT_ALLOW_AUX_TL_RSC;
+                             UCP_EP_INIT_AM_ALLOW_AUX_TL_RSC |
+                             UCP_EP_INIT_KA_ALLOW_AUX_TL_RSC;
```
Comment (brminich) [inline src/ucp/wireup/wireup.c:735]: do we really need aux KA lane here?

File: src/ucp/wireup/wireup_cm.c [hunk 6]
```diff
@@ -423,13 +423,17 @@ static ucs_status_t ucp_cm_ep_init_lanes(ucp_ep_h ep,
...
+    ucp_ep_h ep                         = arg;
+    ucp_worker_h worker                 = ep->worker;
+    ucp_context_h context               = worker->context;
+    ucp_wireup_ep_t *cm_wireup_ep       = ucp_ep_get_cm_wireup_ep(ep);
+    void *ucp_addr                      = NULL; /* Set to NULL to call ucs_free
+                                                   safely */
+    unsigned ep_init_flags              = 0;
+    unsigned ep_init_flags_variants[]   = {
+        UCP_EP_INIT_KA_ALLOW_AUX_TL_RSC, 0, UCP_EP_INIT_CREATE_AM_LANE_ONLY
+    };
```
Comment (brminich) [inline src/ucp/wireup/wireup_cm.c:435]: pls, remove extra whitespaces

File: src/ucp/wireup/wireup_cm.c [hunk 7]
```diff
@@ -443,50 +447,54 @@ static unsigned ucp_cm_client_uct_connect_progress(void *arg)
...
         }
 
-        goto try_fallback;
-    } else if (status != UCS_OK) {
-        goto err;
+        /* Replay pending requests from the tmp_pending_queue */
+        ucp_wireup_replay_pending_requests(ep, &tmp_pending_queue);
+
+        can_fallback =
+                (i != ucs_static_array_size(ep_init_flags_variants)) ||
```
Comment (brminich) [inline src/ucp/wireup/wireup_cm.c:481]: when can it be false?

File: src/ucp/wireup/wireup_cm.c [hunk 8]
```diff
@@ -443,50 +447,54 @@ static unsigned ucp_cm_client_uct_connect_progress(void *arg)
...
+                          &priv_data, &priv_data_length, ep_init_flags);
+        if (status == UCS_ERR_BUFFER_TOO_SMALL) {
+            if (i != ucs_static_array_size(ep_init_flags_variants)) {
+                continue;    
+            }
+
+            goto try_fallback;
+        } else if (status != UCS_OK) {
+            goto err;
+        }
```
Comment (brminich) [inline src/ucp/wireup/wireup_cm.c:495]: maybe better check `status` after the `for` loop?

File: src/ucp/wireup/select.c [hunk 9]
```diff
@@ -1118,6 +1117,8 @@ ucp_wireup_add_am_lane(const ucp_wireup_select_params_t *select_params,
 {
     ucp_worker_h worker            = select_params->ep->worker;
     ucp_tl_bitmap_t tl_bitmap      = select_params->tl_bitmap;
+    unsigned ep_init_flags               = ucp_wireup_ep_init_flags(
+                                                   select_params, select_ctx);
```
Comment (brminich) [inline src/ucp/wireup/select.c:1121]: indentation

Comment (dmitrygx) [reply src/ucp/wireup/select.c]: fixed

File: src/ucp/wireup/wireup_cm.c [hunk 10]
```diff
@@ -423,72 +423,82 @@ static ucs_status_t ucp_cm_ep_init_lanes(ucp_ep_h ep,
...
+
+        break;
     }
 
+    ucs_assert(status == UCS_OK);
```
Comment (brminich) [inline src/ucp/wireup/wireup_cm.c:500]: sorry I meant, checking for `UCS_ERR_BUFFER_TOO_SMALL` here. But looks like it will not make the code clearer. Plz remove this assert, it's not really needed

**Review by brminich — APPROVED** [review]

File: src/ucp/core/ucp_ep.h [hunk 11]
```diff
@@ -160,8 +160,10 @@ enum {
                                                            support CONNECT_TO_IFACE
                                                            mode only */
     UCP_EP_INIT_CREATE_AM_LANE_ONLY    = UCS_BIT(8),  /**< Endpoint requires an AM lane only */
-    UCP_EP_INIT_ALLOW_AUX_TL_RSC       = UCS_BIT(9)   /**< Endpoint allows selecting of auxiliary
-                                                           transports for its configuration */
+    UCP_EP_INIT_KA_ALLOW_AUX_TL_RSC    = UCS_BIT(9),  /**< Endpoint allows selecting of auxiliary
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.h:163]: maybe
UCP_EP_INIT_ALLOW_KA_AUX_TL
UCP_EP_INIT_ALLOW_AM_AUX_TL

File: src/ucp/wireup/wireup_cm.c [hunk 12]
```diff
@@ -423,13 +423,17 @@ static ucs_status_t ucp_cm_ep_init_lanes(ucp_ep_h ep,
...
-                                             safely */
-    unsigned ep_init_flags        = 0;
+    ucp_ep_h ep                       = arg;
+    ucp_worker_h worker               = ep->worker;
+    ucp_context_h context             = worker->context;
+    ucp_wireup_ep_t *cm_wireup_ep     = ucp_ep_get_cm_wireup_ep(ep);
+    void *ucp_addr                    = NULL; /* Set to NULL to call ucs_free
+                                                 safely */
+    unsigned ep_init_flags            = 0;
+    unsigned ep_init_flags_variants[] = {
```
Comment (yosefe) [inline src/ucp/wireup/wireup_cm.c:433]: static const unsigned ep_init_flags_prio

File: src/ucp/wireup/wireup_cm.c [hunk 13]
```diff
@@ -423,13 +423,17 @@ static ucs_status_t ucp_cm_ep_init_lanes(ucp_ep_h ep,
...
-    void *ucp_addr                = NULL; /* Set to NULL to call ucs_free
-                                             safely */
-    unsigned ep_init_flags        = 0;
+    ucp_ep_h ep                       = arg;
+    ucp_worker_h worker               = ep->worker;
+    ucp_context_h context             = worker->context;
+    ucp_wireup_ep_t *cm_wireup_ep     = ucp_ep_get_cm_wireup_ep(ep);
+    void *ucp_addr                    = NULL; /* Set to NULL to call ucs_free
+                                                 safely */
+    unsigned ep_init_flags            = 0;
```
Comment (yosefe) [inline src/ucp/wireup/wireup_cm.c:432]: can we not init to 0 now?

File: src/ucp/wireup/wireup_cm.c [hunk 14]
```diff
@@ -443,50 +447,57 @@ static unsigned ucp_cm_client_uct_connect_progress(void *arg)
...
         }
 
-        goto try_fallback;
-    } else if (status != UCS_OK) {
-        goto err;
+        /* Replay pending requests from the tmp_pending_queue */
+        ucp_wireup_replay_pending_requests(ep, &tmp_pending_queue);
+
+        can_fallback =
+                (i == (ucs_static_array_size(ep_init_flags_variants) - 1)) ||
```
Comment (yosefe) [inline src/ucp/wireup/wireup_cm.c:484]: shouldn't this be <=  (or < size) instead of == ?

File: src/ucp/core/ucp_ep.h (hunk 11 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_ep.h:163]: done

File: src/ucp/wireup/wireup_cm.c (hunk 12 above)
Comment (dmitrygx) [reply src/ucp/wireup/wireup_cm.c:433]: done

File: src/ucp/wireup/wireup_cm.c (hunk 13 above)
Comment (dmitrygx) [reply src/ucp/wireup/wireup_cm.c:432]: done

File: src/ucp/wireup/wireup_cm.c (hunk 14 above)
Comment (dmitrygx) [reply src/ucp/wireup/wireup_cm.c:484]: fixed

File: src/ucp/wireup/wireup_cm.c [hunk 15]
```diff
@@ -443,50 +447,57 @@ static unsigned ucp_cm_client_uct_connect_progress(void *arg)
...
+        status       = ucp_cm_ep_priv_data_pack(
+                          ep, &tl_bitmap, can_fallback,
+                          context->config.ext.sa_client_min_hdr_version,
+                          &priv_data, &priv_data_length, ep_init_flags);
+        if (status == UCS_ERR_BUFFER_TOO_SMALL) {
+            if (i == (ucs_static_array_size(ep_init_flags_variants) - 1)) {
+                continue;
+            }
+
+            goto try_fallback;
```
[the remark this answers is not in the document]
Comment (dmitrygx) [reply src/ucp/wireup/wireup_cm.c:495]: done

File: src/ucp/wireup/wireup_cm.c [hunk 16]
```diff
@@ -443,50 +448,58 @@ static unsigned ucp_cm_client_uct_connect_progress(void *arg)
...
+            continue;
+        } else if (status != UCS_OK) {
+            goto err;
+        }
+
+        break;
+    }
+
+    if ((status == UCS_ERR_BUFFER_TOO_SMALL) &&
+        (i == ucs_static_array_size(ep_init_flags_variants))) {
```
Comment (brminich) [inline src/ucp/wireup/wireup_cm.c:501]: why need to check `i`?

File: src/ucp/wireup/wireup_cm.c [hunk 17]
```diff
@@ -423,16 +423,17 @@ static ucs_status_t ucp_cm_ep_init_lanes(ucp_ep_h ep,
...
-    ucp_ep_h ep                       = arg;
-    ucp_worker_h worker               = ep->worker;
-    ucp_context_h context             = worker->context;
-    ucp_wireup_ep_t *cm_wireup_ep     = ucp_ep_get_cm_wireup_ep(ep);
-    void *ucp_addr                    = NULL; /* Set to NULL to call ucs_free
-                                                 safely */
-    unsigned ep_init_flags            = 0;
-    unsigned ep_init_flags_variants[] = {
-        UCP_EP_INIT_KA_ALLOW_AUX_TL_RSC, 0, UCP_EP_INIT_CREATE_AM_LANE_ONLY
+    static const unsigned ep_init_flags_variants[] = {
```
Comment (yosefe) [inline src/ucp/wireup/wireup_cm.c:426]: IMO should be called  "ep_init_flags_prio" because it's shorter, and we it means it's order of priority for connection establishment

File: src/ucp/wireup/wireup_cm.c (hunk 16 above)
Comment (yosefe) [reply src/ucp/wireup/wireup_cm.c:501]: yeah, maybe just check if (status != UCS_OK)

File: src/ucp/wireup/wireup_cm.c [hunk 18]
```diff
@@ -481,25 +482,26 @@ static unsigned ucp_cm_client_uct_connect_progress(void *arg)
...
                           ep, &tl_bitmap, can_fallback,
                           context->config.ext.sa_client_min_hdr_version,
                           &priv_data, &priv_data_length, ep_init_flags);
         if (status == UCS_ERR_BUFFER_TOO_SMALL) {
-            if (i == (ucs_static_array_size(ep_init_flags_variants) - 1)) {
-                continue;
-            }
-
-            goto try_fallback;
+            continue;
```
Comment (yosefe) [inline src/ucp/wireup/wireup_cm.c:492]: maybe smth like
```c
    status = ...
    if (status == UCS_OK) {
        break;
    } else if (status != UCS_ERR_BUFFER_TOO_SMALL) {
        goto err;
    }
}

File: src/ucp/wireup/wireup_cm.c (hunk 17 above)
Comment (dmitrygx) [reply src/ucp/wireup/wireup_cm.c]: done

File: src/ucp/wireup/wireup_cm.c (hunk 18 above)
Comment (dmitrygx) [reply src/ucp/wireup/wireup_cm.c]: done

File: src/ucp/wireup/wireup_cm.c (hunk 16 above)
Comment (dmitrygx) [reply src/ucp/wireup/wireup_cm.c]: done

**Review by yosefe — APPROVED** [review]

**Review by brminich — APPROVED** [review]

**Comment by changchengx** [comment]
@avildema When testing the DC hash RoCE LAG in io_demo, we need this PR. We can wait for this PR to be merged and rebased to integration4 branch OR mannually apply this PR.
